### Table of Contents

# 1. Import Data

## 1.1 Import the needed libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn as sk
from datetime import date

%matplotlib inline
pd.set_option('display.max_columns', None)

from sklearn.preprocessing import MinMaxScaler
from sklearn.impute import KNNImputer

## 1.2 Import a integrate data

In [2]:
df_crm = pd.read_csv('crm.csv')
df_mkt = pd.read_csv('mkt.csv')
df_sales = pd.read_excel('sales.xlsx')

In [3]:
df = pd.merge(pd.merge(df_crm,df_sales,on='CustomerID',how="inner"),df_mkt,on="CustomerID",how="inner")

## 1.3 Set Index


In [4]:
df.set_index('CustomerID',inplace = True)

## 1.4 Check and removing duplicates

In [5]:
df[df.duplicated()] # checking duplicates

Name  Birthyear   Education Marital_Status  \
CustomerID                                                                   
32924               Mrs. Wanda Terry       1986  Graduation       Divorced   
8276               Mrs. Mary Jackson       1968  HighSchool        Married   
32966               Mr. Steven Baker       1961  HighSchool        Married   
5627              Mr. Jonathan Clark       1988  Graduation        Married   
29357              Mr. Eric Springer       1972         PhD        Married   
26891                  Mr. Sam Peake       1992  Graduation        Married   
21269              Mr. Nathan Glover       1971  Graduation        Married   
32777                  Mr. Joe James       1967      Master        Married   
26285           Mr. Oliver Sanderson       1986       Basic       Together   
5540              Mr. Justin MacLeod       1995  Graduation       Together   
17069             Miss Zoe Underwood       1977  Graduation       Together   
12320         Mr. Christian Morrison       1985      Master        Married   
24260                Mr. Dan Langdon       1993  Graduation         Single   
6470                Miss Diane White       1980  Graduation         Single   
10265                 Mr. Max Martin       1988  HighSchool       Together   
21767                 Mrs. Jan Sharp       1977      Master        Married   
15905              Mr. Charles James       1957         PhD       Together   
28688             Mr. Anthony Gibson       1976  Graduation       Together   
28739          Mr. Stephen Henderson       1970  Graduation        Married   
25052                Mr. Piers Nolan       1977      Master       Divorced   
7586                Mr. Jason Powell       1985      Master        Married   
32495             Miss Wanda Wallace       1962  Graduation       Together   
19295              Mr. Justin Martin       1998  Graduation       Together   
14942              Mr. Stewart North       1978         PhD       Together   
20747               Mr. Luke Lambert       1988         PhD        Married   
12035              Mrs. Heather Hart       1985  Graduation        Married   
8447               Mrs. Grace Davies       1967  Graduation        Married   
25685                Mr. Austin Knox       1955         PhD          Widow   
28625                Mr. Frank Piper       1969       Basic        Married   
28844                  Mr. Neil Dowd       1981  Graduation        Married   
30464       Miss Gabrielle Churchill       1960  Graduation       Together   

              Income  Kid_Younger6  Children_6to18       Date_Adherence  \
CustomerID                                                                
32924        41264.0             1               1  2020-10-03 00:00:00   
8276        118238.0             0               0  2020-10-13 00:00:00   
32966        84974.0             0               1  2022-02-17 00:00:00   
5627         40975.0             1               0  2022-04-28 00:00:00   
29357        77778.0             0               1  2022-04-19 00:00:00   
26891        74147.0             1               0  2022-03-21 00:00:00   
21269        91241.0             0               1  2022-03-28 00:00:00   
32777       101251.0             0               1  2021-12-30 00:00:00   
26285        22345.0             1               0  2021-03-21 00:00:00   
5540         44247.0             1               0  2021-03-08 00:00:00   
17069        75226.0             1               1  2020-11-24 00:00:00   
12320        39747.0             1               0  2021-04-06 00:00:00   
24260       116645.0             0               0  2021-05-18 00:00:00   
6470        127255.0             0               0  2021-06-17 00:00:00   
10265        27838.0             1               0  2021-10-08 00:00:00   
21767        54507.0             1               1  2020-10-27 00:00:00   
15905        95655.0             0               0  2020-09-04 00:00:00   
28688        77992.0            

In [6]:
df = df[~df.duplicated()] # drop duplicates rows

# 2. Explore Data

## 2.0 Data profiling

Se não quiserem instalar a biblioteca não corram esta secção. Caso contrário o comando para instalar é pip install ydata-profiling. No final **apagar esta secção**

In [7]:
#from ydata_profiling import ProfileReport
#profile= ProfileReport (df, title= "DSML_Project")

In [8]:
#profile.to_file('DSML_profile.html')

## 2.1 Basic Exploration

Q: _To check the number of columns and rows_ we used `shape` _attribute_

In [9]:
df.shape

(7000, 26)

> A: _The dataset has **7000 rows** and **26 columns**_

__*Q*__: Check the name of the features of the dataset we used `columns` _attribute_

In [10]:
df.columns

Index(['Name', 'Birthyear', 'Education', 'Marital_Status', 'Income',
       'Kid_Younger6', 'Children_6to18', 'Date_Adherence', 'Recency',
       'MntMeat&Fish', 'MntEntries', 'MntVegan&Vegetarian', 'MntDrinks',
       'MntDesserts', 'MntAdditionalRequests', 'NumOfferPurchases',
       'NumAppPurchases', 'NumTakeAwayPurchases', 'NumStorePurchases',
       'NumAppVisitsMonth', 'Complain', 'Response_Cmp1', 'Response_Cmp2',
       'Response_Cmp3', 'Response_Cmp4', 'Response_Cmp5'],
      dtype='object')

> A: The dataset has the following columns/features names: <br>
        >Index. CustomerID
        >1. 'Name' <br>
        >2. 'Birthyear'<br>
        >3. 'Education'<br>
        >4. 'Marital_Status'<br>
        >5. 'Income'<br>
        >6. 'Kid_Younger6'<br>
        >7. 'Children_6to18'<br>
        >8. 'Date_Adherence'<br>
        >9. 'Recency'<br>
        >10. 'MntMeat&Fish'<br>
        >11. 'MntEntries'<br>
        >12. 'MntVegan&Vegetarian'<br>
        >13. 'MntDrinks'<br>
        >14. 'MntDesserts'<br>
        >15. 'MntAdditionalRequests'<br>
        >16. 'NumOfferPurchases'<br>
        >17. 'NumAppPurchases'<br>
        >18. 'NumTakeAwayPurchases'<br>
        >19. 'NumStorePurchases'<br>
        >20. 'NumAppVisitsMonth'<br>
        >21. 'Complain'<br>
        >22. 'Response_Cmp1'<br>
        >23. 'Response_Cmp2'<br>
        >24. 'Response_Cmp3'<br>
        >25. 'Response_Cmp4'<br>
        >26. 'Response_Cmp5'<br>

Q: First glance of the dataset using `head` and `tail` methods to check the first and last 5 rows.

In [11]:
df.head(3)

Name  Birthyear   Education Marital_Status  \
CustomerID                                                             
14594          Miss Leah Lewis       1981  Graduation         Single   
19070         Miss Grace Black       1974  Graduation       Together   
29888       Mr. Gavin Morrison       1960         PhD       Together   

              Income  Kid_Younger6  Children_6to18       Date_Adherence  \
CustomerID                                                                
14594       113947.0             0               0  2021-01-09 00:00:00   
19070       114297.0             0               0  2021-04-10 00:00:00   
29888       115476.0             0               0  2022-01-12 00:00:00   

            Recency  MntMeat&Fish  MntEntries  MntVegan&Vegetarian  MntDrinks  \
CustomerID                                                                      
14594          43.0        4640.0        3380               5700.0     2240.0   
19070          74.0        7080.0        3680               7620.0     1840.0   
29888          85.0        4130.0         380               4890.0     3640.0   

            MntDesserts  MntAdditionalRequests  NumOfferPurchases  \
CustomerID                                                          
14594             562.0                  154.0                  2   
19070            1220.0                   46.0                  1   
29888             380.0                   28.0                  1   

            NumAppPurchases  NumTakeAwayPurchases  NumStorePurchases  \
CustomerID                                                             
14594                    10                     7                  7   
19070                     5                     9                 12   
29888                     5                     5                  7   

            NumAppVisitsMonth  Complain  Response_Cmp1  Response_Cmp2  \
CustomerID                                                              
14594                       4         0              0              1   
19070                       2         0              0              0   
29888                       1         0              0              0   

            Response_Cmp3  Response_Cmp4  Response_Cmp5  
CustomerID                                               
14594                   0              0              0  
19070                   0              0              0  
29888                   0              0              0

In [12]:
df.tail(3)

Name  Birthyear   Education Marital_Status  \
CustomerID                                                               
29186       Mr. Thomas Churchill       1967         PhD       Together   
23312         Miss Angela Murray       1989  Graduation         Single   
15581          Miss Diana Taylor       1960  Graduation         Single   

              Income  Kid_Younger6  Children_6to18       Date_Adherence  \
CustomerID                                                                
29186        56039.0             0               1  2021-12-15 00:00:00   
23312        44867.0             1               0  2020-11-12 00:00:00   
15581       124379.0             0               0  2021-02-04 00:00:00   

            Recency  MntMeat&Fish  MntEntries  MntVegan&Vegetarian  MntDrinks  \
CustomerID                                                                      
29186          30.0         540.0           0                135.0        0.0   
23312           8.0          80.0         120                210.0      120.0   
15581          80.0        7280.0         422               1692.0     3760.0   

            MntDesserts  MntAdditionalRequests  NumOfferPurchases  \
CustomerID                                                          
29186               0.0                    5.0                  2   
23312             100.0                   17.0                  1   
15581            1880.0                  235.0                  1   

            NumAppPurchases  NumTakeAwayPurchases  NumStorePurchases  \
CustomerID                                                             
29186                     3                     1                  4   
23312                     3                     1                  3   
15581                     6                    12                 11   

            NumAppVisitsMonth  Complain  Response_Cmp1  Response_Cmp2  \
CustomerID                                                              
29186                       6         0              0              0   
23312                       8         0              0              0   
15581                       2         0              0              0   

            Response_Cmp3  Response_Cmp4  Response_Cmp5  
CustomerID                                               
29186                   0              0              0  
23312                   0              0              0  
15581                   1              1              0

Q: To check the basic information of the dataset we've used the `info` method

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7000 entries, 14594 to 15581
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Name                   7000 non-null   object 
 1   Birthyear              7000 non-null   int64  
 2   Education              6986 non-null   object 
 3   Marital_Status         7000 non-null   object 
 4   Income                 7000 non-null   float64
 5   Kid_Younger6           7000 non-null   int64  
 6   Children_6to18         7000 non-null   int64  
 7   Date_Adherence         7000 non-null   object 
 8   Recency                6977 non-null   float64
 9   MntMeat&Fish           7000 non-null   float64
 10  MntEntries             7000 non-null   int64  
 11  MntVegan&Vegetarian    7000 non-null   float64
 12  MntDrinks              6972 non-null   float64
 13  MntDesserts            7000 non-null   float64
 14  MntAdditionalRequests  7000 non-null   float64
 15 

>A: We can observe the data type of the dataset and the how many of features per data type  `dtypes: float64 - (7), int64 - (15), object - (4)`, the memory usage of `1.4+MB`, and the non-null values present per columns. <br>
> Using only `info` method we understand that `'Education', 'Recency', 'MntDrinks'` have __14, 23, 28 null values__ that require some action.

# 2.2 Statistical Exploration

## 2.2.1 Numerical Variables

In [14]:
df.describe()

Birthyear         Income  Kid_Younger6  Children_6to18      Recency  \
count  7000.000000    7000.000000   7000.000000     7000.000000  6977.000000   
mean   1976.451429   77988.962407      0.440571        0.490571    49.235058   
std      11.996270   35409.810253      0.543477        0.542174    28.922688   
min    1948.000000    2493.800000      0.000000        0.000000     0.000000   
25%    1967.000000   51586.250000      0.000000        0.000000    24.000000   
50%    1977.000000   77190.000000      0.000000        0.000000    49.000000   
75%    1985.000000  102016.250000      1.000000        1.000000    74.000000   
max    2005.000000  237639.725000      2.000000        2.000000    99.000000   

       MntMeat&Fish   MntEntries  MntVegan&Vegetarian    MntDrinks  \
count   7000.000000  7000.000000          7000.000000  6972.000000   
mean    3079.523800   534.749429          2785.050786   545.657544   
std     3370.377166   787.846684          3908.718244   805.149088   
min        0.000000     0.000000             0.000000     0.000000   
25%      250.000000    40.000000           240.000000    40.000000   
50%     1820.000000   180.000000          1110.000000   180.000000   
75%     5070.000000   680.000000          3795.000000   700.000000   
max    14980.000000  3980.000000         25974.000000  3980.000000   

       MntDesserts  MntAdditionalRequests  NumOfferPurchases  NumAppPurchases  \
count  7000.000000            7000.000000        7000.000000      7000.000000   
mean    540.656029              42.556186           2.448429         6.015714   
std     802.221866              49.650747           2.306968         2.745537   
min       0.000000               0.000000           0.000000         0.000000   
25%      40.000000               9.000000           1.000000         4.000000   
50%     180.000000              24.000000           2.000000         6.000000   
75%     680.000000              57.000000           3.000000         8.000000   
max    3980.000000             249.000000          16.000000        13.000000   

       NumTakeAwayPurchases  NumStorePurchases  NumAppVisitsMonth  \
count           7000.000000        7000.000000        7000.000000   
mean               3.834571           5.790571           5.278286   
std                3.331142           3.295708           2.748596   
min                0.000000           0.000000           0.000000   
25%                1.000000           3.000000           3.000000   
50%                3.000000           5.000000           5.000000   
75%                5.000000           8.000000           7.000000   
max               24.000000          13.000000          20.000000   

          Complain  Response_Cmp1  Response_Cmp2  Response_Cmp3  \
count  7000.000000    7000.000000    7000.000000    7000.000000   
mean      0.010286       0.079143       0.073286       0.083000   
std       0.100903       0.269981       0.260624       0.275902   
min       0.000000       0.000000       0.000000       0.000000   
25%       0.000000       0.000000       0.000000       0.000000   
50%       0.000000       0.000000       0.000000       0.000000   
75%       0.000000       0.000000       0.000000       0.000000   
max       1.000000       1.000000       1.000000       1.000000   

       Response_Cmp4  Response_Cmp5  
count    7000.000000    7000.000000  
mean        0.065857       0.014286  
std         0.248050       0.118675  
min         0.000000       0.000000  
25%         0.000000       0.000000  
50%         0.000000       0.000000  
75%         0.000000       0.000000  
max         1.000000       1.000000

> _The describe return we can get a first glance and make some conclusion:_

>__Birthyear__ - could originate an Age column for readability purposes<br>
__Income__ - Min and Max are very far from each other and far from the mean value which could indicate outliers<br>
__Recency__ - 6977 valid values, hence we should look in deep and decide on how to minimize that effect of missing values<br>
__MntMeat&Fish__ - Min and Max are distant from each other and have high standard deviation which could effect some future conclusion<br>
__MntEntries__ - Again has high standard deviation that we should analyze, Min and Max far apart, similar to MntMeat&Fish<br>
__MntVegan&Vegetarian__ - Similar to the previous two Mnt columns<br>
__MntDrinks, MntDesserts__ - Seems to be very similar between them<br>
__MntAdditionalRequests__ - The max value standard deviation seems high and also the max value very far apart from the mean<br>
__NumOfferPurchases, NumTakeAwayPurchases, NumAppVisitsMonth__  - Have a max value to distante from the mean that could be true but we need to take into account<br>
__NumAppPurchases, SumStorePurchases__ - Seems does not have strange summary statistcs<br>
__Kid_Younger6, Children_6to18__ - 75% of clients have at least one child

**Q**: Skewness of each variable 

In [15]:
df.skew()

C:\Users\guisi\AppData\Local\Temp\ipykernel_10888\547062910.py:1: FutureWarning: The default value of numeric_only in DataFrame.skew is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.skew()


Birthyear               -0.092478
Income                   0.844721
Kid_Younger6             0.693587
Children_6to18           0.451693
Recency                  0.017676
MntMeat&Fish             1.148862
MntEntries               2.087220
MntVegan&Vegetarian      2.486897
MntDrinks                2.046308
MntDesserts              2.057811
MntAdditionalRequests    1.826522
NumOfferPurchases        2.860514
NumAppPurchases          0.525873
NumTakeAwayPurchases     2.254158
NumStorePurchases        0.623254
NumAppVisitsMonth        1.005192
Complain                 9.709429
Response_Cmp1            3.118572
Response_Cmp2            3.275500
Response_Cmp3            3.023677
Response_Cmp4            3.501448
Response_Cmp5            8.187993
dtype: float64

Concerning the variables' skewness, we can conclude the following:
- `Moderate skewness (between |0.5| and |1.0|)`: Birthyear, Income, Kid_Younger6, Children_6to18, Recency, NumAppPurchases, NumStorePurchases, NumAppVisitsMonth
- `High skewness (higher than |1.0|)`: MntMeat&Fish, MntEntries, MntVegan&Vegetarian, MntDrinks, MntDesserts, MntAdditionalRequests, NumOfferPurchases, NumTakeAwayPurchases, Complain, Response_Cmp1, Response_Cmp2, Response_Cmp3, Response_Cmp4, Response_Cmp5

In [16]:
df.kurt()

C:\Users\guisi\AppData\Local\Temp\ipykernel_10888\3952544226.py:1: FutureWarning: The default value of numeric_only in DataFrame.kurt is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.kurt()


Birthyear                -0.857341
Income                    2.649035
Kid_Younger6             -0.651127
Children_6to18           -0.973333
Recency                  -1.197978
MntMeat&Fish              0.532098
MntEntries                4.096167
MntVegan&Vegetarian       8.431538
MntDrinks                 3.839256
MntDesserts               3.813464
MntAdditionalRequests     3.084190
NumOfferPurchases        10.997946
NumAppPurchases          -0.222899
NumTakeAwayPurchases      8.582138
NumStorePurchases        -0.694327
NumAppVisitsMonth         4.992870
Complain                 92.299387
Response_Cmp1             7.727700
Response_Cmp2             8.731398
Response_Cmp3             7.144665
Response_Cmp4            10.263068
Response_Cmp5            65.061813
dtype: float64

Features with kurtosis higher than 3 could indicate presence of outliers, hence we should have special considerantion with the following features:
>MntEntries, MntVegan&Vegetarian, MntDrinks, MntDesserts, NumOfferPurchases, NumAppVisitsMonth

Note: Binomial Variables Complain, and Response_Cmp1 the kurtosis we will not consider as outliers

## 2.2.2 Categorical Values

In [17]:
df.describe(include = object)

Name   Education Marital_Status       Date_Adherence
count                7000        6986           7000                 7000
unique               6241           9             10                  701
top     Mr. Stewart Grant  Graduation        Married  2020-09-19 00:00:00
freq                    3        3497           2830                   23

> We can conclude that the education as **14 missing** values

#### Level/Possible values of Categorical Features

### `Name` prefix unique values and count

In [18]:
df['Name'].str.partition(" ")[0].value_counts()

Mr.     5133
Miss    1469
Mrs.     398
Name: 0, dtype: int64

With the prefix we can generate a `gender` feature to further explore the dataset. We will deal with that in the data transformation capther

#### **`Gender`** feature creation

In [19]:
df["Gender"] = df['Name'].str.partition(" ")[0]
df = df.replace({"Gender":{"Mr.": 1,"Miss": 0,"Mrs.": 0}})

### `Education` unique values and count

In [20]:
df["Education"].value_counts()

Graduation    3497
PhD           1494
Master        1135
HighSchool     663
Basic          179
master           7
graduation       7
phd              2
highschool       2
Name: Education, dtype: int64

We have some issues that will need trasformatioin:<br>
- Graduation, Master, HighSchool are written in different ways<br>
- `Basic` and `HighSchool` need different levels?

#### Education standardization

In [21]:
df = df.replace({"Education":{"master":"Master", "graduation":"Graduation", "phd":"PhD","highschool":"HighSchool"}})

### `Marital_Status` unique values and count

In [22]:
df["Marital_Status"].value_counts()

Married     2830
Together    1683
Single      1525
Divorced     637
Widow        243
married       36
together      23
single        13
divorced       8
widow          2
Name: Marital_Status, dtype: int64

Similarly to previous feature we also have some issues that need transformation:<br>
- Married, Together, Single, Divorced and Widow are written with lower and capital letters
- We could also consider that Married and Together are similar and joined them in the same level<br>

#### Marital_Status standardization

In [23]:
df = df.replace({"Marital_Status":{"married":"Married", "together":"Married", "single":"Single","widow":"Widow","divorced":"Divorced","Together":"Married"}})
df["Marital_Status"].value_counts()

Married     4572
Single      1538
Divorced     645
Widow        245
Name: Marital_Status, dtype: int64

`Date_Adherence` unqiue values and count

In [24]:
df["Date_Adherence"].value_counts()

2020-09-19    23
2021-08-22    20
2021-07-30    20
2022-05-23    20
2021-08-21    19
              ..
2021-03-09     3
2020-09-18     2
2020-07-30     2
2020-08-29     2
2021-03-22     2
Name: Date_Adherence, Length: 701, dtype: int64

`Date_Adherence` is a date and will need transformation to a date format for further exploration

## 2.3 Visual Exploration

### 2.3.1 Numerical Variables

## 2.4. In-Depth Exploration

# 3. Preprocess Data

## 3.1. Data Cleaning

### 3.1.1. Outliers

- boxplot?

### 3.1.2. Missing Values

In [25]:
df["Response_Cmp1"].isna().sum()

0

In [26]:
df["Response_Cmp2"].isna().sum()

0

In [27]:
df["Response_Cmp3"].isna().sum()

0

In [28]:
df["Response_Cmp4"].isna().sum()


0

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7000 entries, 14594 to 15581
Data columns (total 27 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Name                   7000 non-null   object 
 1   Birthyear              7000 non-null   int64  
 2   Education              6986 non-null   object 
 3   Marital_Status         7000 non-null   object 
 4   Income                 7000 non-null   float64
 5   Kid_Younger6           7000 non-null   int64  
 6   Children_6to18         7000 non-null   int64  
 7   Date_Adherence         7000 non-null   object 
 8   Recency                6977 non-null   float64
 9   MntMeat&Fish           7000 non-null   float64
 10  MntEntries             7000 non-null   int64  
 11  MntVegan&Vegetarian    7000 non-null   float64
 12  MntDrinks              6972 non-null   float64
 13  MntDesserts            7000 non-null   float64
 14  MntAdditionalRequests  7000 non-null   float64
 15 

In [30]:
df.isna().sum()

Name                      0
Birthyear                 0
Education                14
Marital_Status            0
Income                    0
Kid_Younger6              0
Children_6to18            0
Date_Adherence            0
Recency                  23
MntMeat&Fish              0
MntEntries                0
MntVegan&Vegetarian       0
MntDrinks                28
MntDesserts               0
MntAdditionalRequests     0
NumOfferPurchases         0
NumAppPurchases           0
NumTakeAwayPurchases      0
NumStorePurchases         0
NumAppVisitsMonth         0
Complain                  0
Response_Cmp1             0
Response_Cmp2             0
Response_Cmp3             0
Response_Cmp4             0
Response_Cmp5             0
Gender                    0
dtype: int64

- **`Education`**, **`Recency`**, **`MntDrinks`** and **`MntTotal`** (due to dependancy of `MntDrinks`) have missing values

#### Filling the missing values

Fill `Education` with the mode

In [31]:
df["Education"].fillna(df["Education"].mode()[0], inplace = True)

Fill `Recency` with the median value

In [32]:
df["Recency"].fillna(df["Recency"].mean(), inplace = True)

In [33]:
#df.drop(columns = "MntTotal", inplace = True)

In [34]:
df_mnt = df[[ 'MntMeat&Fish', 'MntEntries', 'MntVegan&Vegetarian', 'MntDrinks',
       'MntDesserts', 'MntAdditionalRequests']]

imputer = KNNImputer(n_neighbors=3)
array_impute = imputer.fit_transform(df_mnt)
df_mnt = pd.DataFrame(array_impute, columns = df_mnt.columns)

In [35]:
df["MntDrinks"] = df_mnt["MntDrinks"].values

In [36]:
df["MntTotal"] = df['MntMeat&Fish'] + df['MntEntries'] + df['MntVegan&Vegetarian'] + df['MntDrinks'] + df['MntDesserts']

In [37]:
df.isna().sum()

Name                     0
Birthyear                0
Education                0
Marital_Status           0
Income                   0
Kid_Younger6             0
Children_6to18           0
Date_Adherence           0
Recency                  0
MntMeat&Fish             0
MntEntries               0
MntVegan&Vegetarian      0
MntDrinks                0
MntDesserts              0
MntAdditionalRequests    0
NumOfferPurchases        0
NumAppPurchases          0
NumTakeAwayPurchases     0
NumStorePurchases        0
NumAppVisitsMonth        0
Complain                 0
Response_Cmp1            0
Response_Cmp2            0
Response_Cmp3            0
Response_Cmp4            0
Response_Cmp5            0
Gender                   0
MntTotal                 0
dtype: int64

## 3.2. Data Transformation

### 3.2.1. Create new Variables

### Utils

#### Creating Age variable from the Birthyear

In [38]:
df['Age'] = df.Birthyear.apply(lambda x: date.today().year-x)

In [39]:
#### Creating card adherence age variable from the Date adherence

In [40]:
from datetime import datetime
df = df.replace({"Date_Adherence":{"2/29/2022": datetime.strptime("2022-03-01", '%Y-%m-%d')}}) #2022 is not a leap year, therefore 29/02/2022 is not a possible day

In [41]:
df['daysAsCardClient'] = df['Date_Adherence'].apply(lambda x: (date.today() - x.date()).days)

In [42]:
edu_encode = pd.get_dummies(df.Education, drop_first= True)
df = pd.concat([df, edu_encode], axis = 1)
df.drop('Education', axis = 1, inplace = True)

In [43]:
marital_encode = pd.get_dummies(df.Marital_Status, drop_first= True)
df = pd.concat([df, marital_encode], axis = 1)
df.drop('Marital_Status', axis= 1, inplace = True)

In [44]:
df["MntTotal"] = df['MntMeat&Fish'] + df['MntEntries'] + df['MntVegan&Vegetarian'] + df['MntDrinks'] + df['MntDesserts'] + df['MntAdditionalRequests']
df["MntTotal"]
# em falta Mnt Add Requests

CustomerID
14594    16676.0
19070    21486.0
29888    13448.0
12221     5634.0
21107    17303.0
          ...   
30686    15665.0
17210    24246.0
29186      680.0
23312      647.0
15581    15269.0
Name: MntTotal, Length: 7000, dtype: float64

In [45]:
df['Mnt_pday_card']= df.MntTotal/df.daysAsCardClient

In [46]:
df['Response_Campaigns'] = df['Response_Cmp1'] + df['Response_Cmp2'] + df['Response_Cmp3'] + df['Response_Cmp4'] + df[
       'Response_Cmp5']

## Data Review

Ver a dataframe no seu estado final
Drop: Id, name, birthyear, date_adherence, total_kids, mntTotal

In [47]:
df.head()

Name  Birthyear    Income  Kid_Younger6  \
CustomerID                                                          
14594          Miss Leah Lewis       1981  113947.0             0   
19070         Miss Grace Black       1974  114297.0             0   
29888       Mr. Gavin Morrison       1960  115476.0             0   
12221          Mr. Ryan Glover       1980  101172.0             1   
21107        Mr. Paul Anderson       1971  123128.0             0   

            Children_6to18 Date_Adherence  Recency  MntMeat&Fish  MntEntries  \
CustomerID                                                                     
14594                    0     2021-01-09     43.0        4640.0        3380   
19070                    0     2021-04-10     74.0        7080.0        3680   
29888                    0     2022-01-12     85.0        4130.0         380   
12221                    1     2021-10-12     25.0        3590.0           0   
21107                    0     2020-12-11     64.0        5390.0        1126   

            MntVegan&Vegetarian  MntDrinks  MntDesserts  \
CustomerID                                                
14594                    5700.0     2240.0        562.0   
19070                    7620.0     1840.0       1220.0   
29888                    4890.0     3640.0        380.0   
12221                    1770.0       80.0        180.0   
21107                    9495.0      656.0        562.0   

            MntAdditionalRequests  NumOfferPurchases  NumAppPurchases  \
CustomerID                                                              
14594                       154.0                  2               10   
19070                        46.0                  1                5   
29888                        28.0                  1                5   
12221                        14.0                  5                8   
21107                        74.0                  0                4   

            NumTakeAwayPurchases  NumStorePurchases  NumAppVisitsMonth  \
CustomerID                                                               
14594                          7                  7                  4   
19070                          9                 12                  2   
29888                          5                  7                  1   
12221                          3                  8                  5   
21107                          7                  8                  5   

            Complain  Response_Cmp1  Response_Cmp2  Response_Cmp3  \
CustomerID                                                          
14594              0              0              1              0   
19070              0              0              0              0   
29888              0              0              0              0   
12221              0              0              0              0   
21107              0              1              0              1   

            Response_Cmp4  Response_Cmp5  Gender  MntTotal  Age  \
CustomerID                                                        
14594                   0              0       0   16676.0   42   
19070                   0              0       0   21486.0   49   
29888                   0              0       1   13448.0   63   
12221                   0              0       1    5634.0   43   
21107                   0              0       1   17303.0   52   

            daysAsCardClient  Graduation  HighSchool  Master  PhD  Married  \
CustomerID                                                                   
14594                    823           1           0       0    0        0   
19070                    732           1           0       0    0        1   
29888                    455           0           0       0    1        1   
12221                    547           1           0       0    0        1   
21107                    852           0           1       0    0        0   

            Single  Widow  Mnt_pday

In [48]:
df.describe()

Birthyear         Income  Kid_Younger6  Children_6to18      Recency  \
count  7000.000000    7000.000000   7000.000000     7000.000000  7000.000000   
mean   1976.451429   77988.962407      0.440571        0.490571    49.235058   
std      11.996270   35409.810253      0.543477        0.542174    28.875126   
min    1948.000000    2493.800000      0.000000        0.000000     0.000000   
25%    1967.000000   51586.250000      0.000000        0.000000    24.000000   
50%    1977.000000   77190.000000      0.000000        0.000000    49.000000   
75%    1985.000000  102016.250000      1.000000        1.000000    74.000000   
max    2005.000000  237639.725000      2.000000        2.000000    99.000000   

       MntMeat&Fish   MntEntries  MntVegan&Vegetarian    MntDrinks  \
count   7000.000000  7000.000000          7000.000000  7000.000000   
mean    3079.523800   534.749429          2785.050786   546.463867   
std     3370.377166   787.846684          3908.718244   805.423098   
min        0.000000     0.000000             0.000000     0.000000   
25%      250.000000    40.000000           240.000000    40.000000   
50%     1820.000000   180.000000          1110.000000   180.000000   
75%     5070.000000   680.000000          3795.000000   700.000000   
max    14980.000000  3980.000000         25974.000000  3980.000000   

       MntDesserts  MntAdditionalRequests  NumOfferPurchases  NumAppPurchases  \
count  7000.000000            7000.000000        7000.000000      7000.000000   
mean    540.656029              42.556186           2.448429         6.015714   
std     802.221866              49.650747           2.306968         2.745537   
min       0.000000               0.000000           0.000000         0.000000   
25%      40.000000               9.000000           1.000000         4.000000   
50%     180.000000              24.000000           2.000000         6.000000   
75%     680.000000              57.000000           3.000000         8.000000   
max    3980.000000             249.000000          16.000000        13.000000   

       NumTakeAwayPurchases  NumStorePurchases  NumAppVisitsMonth  \
count           7000.000000        7000.000000        7000.000000   
mean               3.834571           5.790571           5.278286   
std                3.331142           3.295708           2.748596   
min                0.000000           0.000000           0.000000   
25%                1.000000           3.000000           3.000000   
50%                3.000000           5.000000           5.000000   
75%                5.000000           8.000000           7.000000   
max               24.000000          13.000000          20.000000   

          Complain  Response_Cmp1  Response_Cmp2  Response_Cmp3  \
count  7000.000000    7000.000000    7000.000000    7000.000000   
mean      0.010286       0.079143       0.073286       0.083000   
std       0.100903       0.269981       0.260624       0.275902   
min       0.000000       0.000000       0.000000       0.000000   
25%       0.000000       0.000000       0.000000       0.000000   
50%       0.000000       0.000000       0.000000       0.000000   
75%       0.000000       0.000000       0.000000       0.000000   
max       1.000000       1.000000       1.000000       1.000000   

       Response_Cmp4  Response_Cmp5       Gender      MntTotal          Age  \
count    7000.000000    7000.000000  7000.000000   7000.000000  7000.000000   
mean        0.065857       0.014286     0.733286   7529.000095    46.548571   
std         0.248050       0.118675     0.442273   7649.219409    11.996270   
min         0.000000       0.000000     0.000000     16.800000    18.000000   
25%         0.000000       0.000000     0.000000    812.750000    38.000000   
50%         0.000000       0.000000     1.000000   4652.500000    46.000000   
75%         0.000000       0.000000     1.000000  12764.750000    56.000000   
max         1.000000       1.000000     1.000000  33256.000000    75.

In [49]:
df_train= df.copy()
df_train.drop(['Name', 'Birthyear', 'Date_Adherence', 'Response_Cmp1', 'Response_Cmp2', 'Response_Cmp3', 'Response_Cmp4', 'Response_Cmp5',
               'Kid_Younger6', 'Children_6to18'], axis = 1, inplace = True)

In [50]:
df_train.isna().sum()

Income                   0
Recency                  0
MntMeat&Fish             0
MntEntries               0
MntVegan&Vegetarian      0
MntDrinks                0
MntDesserts              0
MntAdditionalRequests    0
NumOfferPurchases        0
NumAppPurchases          0
NumTakeAwayPurchases     0
NumStorePurchases        0
NumAppVisitsMonth        0
Complain                 0
Gender                   0
MntTotal                 0
Age                      0
daysAsCardClient         0
Graduation               0
HighSchool               0
Master                   0
PhD                      0
Married                  0
Single                   0
Widow                    0
Mnt_pday_card            0
Response_Campaigns       0
dtype: int64

# Feature Selection

As the problem is not a classification problem we will need to adapt our data so it can be used in classifiers. We will use Random Forest Classifiers as tools for feature selection, using MntTotal as the target variable, as our goal is to devise a marketing campaign that aims to increase sales

In [81]:
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.feature_selection import SelectFromModel
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [96]:
x= df_train.drop(['MntTotal', 'Mnt_pday_card', 'MntAdditionalRequests', 'MntDesserts', 'MntDrinks', 'MntEntries', 'MntMeat&Fish', 'MntVegan&Vegetarian'] , axis= 1)
y= df_train['MntTotal']
x_train, x_test, y_train, y_test = train_test_split( x, y, test_size=0.2, random_state=1)

y_train = np.array(y_train).reshape(-1, 1)
y_test = np.array(y_test).reshape(-1, 1)

In [97]:
scalerx = MinMaxScaler()
scalery = MinMaxScaler()

scalerx = scalerx.fit(x_train)
scalery = scalery.fit(y_train)

x_train = pd.DataFrame(scalerx.transform(x_train), columns= x.columns)
x_test = pd.DataFrame(scalerx.transform(x_test), columns= x.columns)

y_train = pd.DataFrame(scalery.transform(y_train))
y_test = pd.DataFrame(scalery.transform(y_test))
#df_train.describe()

In [98]:
RFR = RandomForestRegressor(n_estimators=100, max_depth=5)
RFR.fit(x_train, y_train)

C:\Users\guisi\AppData\Local\Temp\ipykernel_10888\2527319989.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  RFR.fit(x_train, y_train)


RandomForestRegressor(max_depth=5)

In [99]:
importances = pd.Series(RFR.feature_importances_, index= x_train.columns)
importances

Income                  0.643063
Recency                 0.001526
NumOfferPurchases       0.003378
NumAppPurchases         0.007597
NumTakeAwayPurchases    0.276164
NumStorePurchases       0.045888
NumAppVisitsMonth       0.001455
Complain                0.000070
Gender                  0.000074
Age                     0.001569
daysAsCardClient        0.016228
Graduation              0.000147
HighSchool              0.000155
Master                  0.000073
PhD                     0.000045
Married                 0.000109
Single                  0.000061
Widow                   0.000496
Response_Campaigns      0.001901
dtype: float64

In [109]:
threshold = importances.median()
selected_features = x_train.reset_index(drop=True).loc[:, importances >= threshold]
print(selected_features.columns.to_series().reset_index(drop=True).to_string(index=False))

              Income
             Recency
   NumOfferPurchases
     NumAppPurchases
NumTakeAwayPurchases
   NumStorePurchases
   NumAppVisitsMonth
                 Age
    daysAsCardClient
  Response_Campaigns


In [101]:
new_RFR = RandomForestRegressor(n_estimators=100, max_depth=5, random_state=1)
new_RFR.fit(selected_features, y_train)

C:\Users\guisi\AppData\Local\Temp\ipykernel_10888\2501353201.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  new_RFR.fit(selected_features, y_train)


RandomForestRegressor(max_depth=5, random_state=1)

In [102]:
x_test= x_test[selected_features.columns]
y_pred = new_RFR.predict(x_test)
mse = mean_squared_error(y_test, y_pred)
print("MSE: ", mse)

MSE:  0.008360180382056603


## Data scaling
min max: income, recency, mnt..., purchases ..., age, daysasClient, mnt per ...

In [ ]:
scaler = MinMaxScaler()
df_train = pd.DataFrame(scaler.fit_transform(df_train))
#df_train.describe()

## PCA


A implementação do pca acima pareceu me estranha. deixo aqui outra e quando reunirmos vemos

In [ ]:
from sklearn.decomposition import PCA
pca = PCA()
df_train2= df_train.copy()
pca.fit(df_train2)
var= pca.explained_variance_ratio_
var1=np.cumsum(np.round(pca.explained_variance_ratio_, decimals=4)*100)

plt.title("PCA Variance against num of Componmnets")
plt.ylabel("Variance %")
plt.xlabel("Number of componments")
l = plt.axhline(80, color="red")

plt.plot(var1)
plt.grid()

In [ ]:
pca = PCA(n_components=7)
pca_train=pca.fit_transform(df_train2)
pca_train

In [ ]:
df_train

7 variaveis explicam ~80% da variancia

## Loading Scores for each PC

##  Correlação entre PC's e as variáveis originais

## Model Train

In [ ]:
from sklearn.cluster import KMeans

kmeans= KMeans(n_clusters = 7, max_iter =10000, random_state= 1)
kmeans.fit(pca_train)
pca_train_label = kmeans.labels_

In [ ]:
kmeans2= KMeans(n_clusters = 7, max_iter = 10000, random_state = 1)
kmeans2.fit(df_train)
df_train_label = kmeans2.labels_

In [ ]:
dfTF= pd.DataFrame()
dfTF['Compare'] = pca_train_label== df_train_label
#pd.DataFrame(dfTF.groupby('Compare'))
dfTF.groupby(['Compare']).size().reset_index(name='counts')

### 3.2.2. Misclassifications

### 3.2.3. Incoherencies

> 3.2.3.1 Clients that spent money but never had a registered Purchase

In [ ]:
df[(df[['MntMeat&Fish', 'MntEntries',
       'MntVegan&Vegetarian', 'MntDrinks', 'MntDesserts',
       'MntAdditionalRequests']].sum(axis = 1) >= 0) & (df[['NumOfferPurchases', 'NumAppPurchases',
       'NumTakeAwayPurchases', 'NumStorePurchases']].sum(axis = 1) <= 0)].shape

### 3.2.4. Binning

### 3.2.5. Reclassification

### 3.2.6. Power Transform

## 3.3. Data Reduction

### 3.3.1. Multicollinearity - Check correlation

### 3.3.2. Unary Variables

### 3.3.3. Variables with a high percentage of missing values

## 3.2. Back to Data Transformation

### 3.2.7. Apply ordinal encoding and create Dummy variables

### 3.2.8. Scaling